File: apps/backend/Models/Transaction.cs
using System.ComponentModel.DataAnnotations;
using DigitalLedger.Api.Models.Enums;

namespace DigitalLedger.Api.Models;

public class Transaction
{
    public Guid Id { get; set; }

    [MaxLength(50)]
    public string? ReferenceNumber { get; set; }

    [Required]
    public required string Description { get; set; }

    [Required]
    public DateTime Date { get; set; }

    [Required]
    public TransactionStatus Status { get; set; }

    public Guid CreatedBy { get; set; }
    
    public ApplicationUser? Creator { get; set; }

    public Guid? ApprovedBy { get; set; }
    
    public ApplicationUser? Approver { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    public List<TransactionEntry> Entries { get; set; } = new();
}

File: apps/frontend/src/types/transaction.ts
export type TransactionStatus = 'DRAFT' | 'PENDING_APPROVAL' | 'POSTED' | 'VOIDED';
export type EntryType = 'DEBIT' | 'CREDIT';

export interface TransactionEntry {
  id: string;
  accountId: string;
  accountCode: string;
  accountName: string;
  amount: number;
  entryType: EntryType;
  description?: string;
}

export interface Transaction {
  id: string;
  referenceNumber?: string;
  description: string;
  date: string; // ISO string form
  status: TransactionStatus;
  createdBy: string;
  creatorName?: string;
  approvedBy?: string;
  approverName?: string;
  createdAt: string;
  updatedAt: string;
  entries: TransactionEntry[];
}

export interface CreateTransactionEntryDto {
  accountId: string;
  amount: number;
  entryType: EntryType;
  description?: string;
}

export interface UpdateTransactionEntryDto {
  id?: string;
  accountId: string;
  amount: number;
  entryType: EntryType;
  description?: string;
}

export interface CreateTransactionDto {
  referenceNumber?: string;
  description: string;
  date: string;
  status: TransactionStatus;
  entries: CreateTransactionEntryDto[];
}

export interface UpdateTransactionDto {
  referenceNumber?: string;
  description: string;
  date: string;
  status: TransactionStatus;
  entries: UpdateTransactionEntryDto[];
}
